### DISCUSSION TITLE
liner install & liner uninstall

### DISCUSSION THREAD

**Discussion by ToBinio** [PaulDance/cargo-liner#18] [Ideas]
It would be nice to have a simple way of adding crates to the `liner.toml` and installing them which does not include needed to edit the `liner.toml` by hand

as long as you do not have any custom config you can always just re-import all the installed crates. But still always having to enter a second command after installing a crate feels a bit sluggish.

it could also be a `add` / `remove` command which edits the `liner.toml` and runs `liner ship` which an extra flag
but this would just be `liner install` with a different cover

I am opening this as a discussion since I feel like it kinda falls under

> Non-goals:
> * Re-implementing half of Cargo for small functionalities.

note:
maybe there is already a simpler way that I just missed

sry if I am a bit annoying by opening all these issues, i have just discovered this project and really love the idea

**Comment by PaulDance** [comment]
Hi!

> sry if I am a bit annoying by opening all these issues

Don't worry about it, that's what they're made for! Feedback is always appreciated and it's not like I'm flooded with it.

> I am opening this as a discussion since I feel like it kinda falls under

Actually, I don't even think so: this is behavior specific to the current project since it is only about managing its configuration and not about Cargo itself. I therefore find this valid.

> It would be nice to have a simple way of adding crates to the liner.toml and installing them which does not include needed to edit the liner.toml by hand

I agree. A simple idea I had in the past was to have a command to spawn an editor on it. It would have simplified the process a tiny bit, but not much more.

The thing I want to avoid is automated, opaque and behind-the-scenes edition of the configuration. However, having new commands dedicated to it would make it more explicit and thus more user-friendly in that aspect. Still, the edition would have to be tested to be as stable as possible, i.e. not reformat every line in order to add only one. It could end up being a big stack of caveats, but at least the interaction would be explicit.

Regarding the design, I'll have to properly think about it, but as a first idea: maybe add these commands as sub-subcommands, for example `liner config add` and `liner config remove`? Not sure it would be the best though.

> i have just discovered this project and really love the idea

:heart:

**Reply by ToBinio** [reply]
>  i.e. not reformat every line in order to add only one

I think this shouldn't be a big problem with [toml_edit](https://crates.io/crates/toml_edit)

there could also be a small diff display where before writing the change it first shows you how the config will look like
I don't have any experience with diffing libraries but they look rather straightforward e.g. [prettydiff](https://crates.io/crates/prettydiff)

**Reply by PaulDance** [reply]
Yes, good ideas!
